### PR TITLE
Fixed test for get_optics

### PR DESCRIPTION
**Pull request by narJH27**

### CODE REVIEW

**Comment by mirceaulinic** [comment]
Rather than changing the base tests because https://github.com/napalm-automation/napalm-eos/pull/32 is failing, why not leave the output as dict and adapt the output from napalm-eos? I find it more elegant to have it indexed by the interface name.

**Comment by narJH27** [comment]
The only issue with adopting that approach is that, IOSXR does not conform to the same parameters. JUNOS does. @dbarrosop your thoughts?

File: napalm_base/test/base.py [hunk 1]
```diff
@@ -473,14 +473,14 @@ def test_get_optics(self):
...
 
-        result = isinstance(get_optics, dict)
+        result = isinstance(get_optics, list)
 
-        for intf, intf_optic in get_optics.iteritems():
-            for phy_chann in intf_optic:
-                for chan in intf_optic[phy_chan]:
-                    for item in intf_optic[phy_chan][chan]:
+        for item in get_optics:
+            for phy_chann in item:
```
Comment (mirceaulinic) [inline napalm_base/test/base.py]: ```phy_chan```

**Comment by mirceaulinic** [comment]
What do you mean? You can't get the interface name? In your test I do not see any variable that might play the role of the interface name. Can you provide a proper example of the output expected?

**Comment by mirceaulinic** [comment]
And what's the XML request on IOS-XR to retrieve the optics details?

File: napalm_base/test/base.py (hunk 1 above)
Comment (narJH27) [inline napalm_base/test/base.py]: yes that is a typo. will be fixed!

**Comment by dbarrosop** [comment]
I would certainly leave the interface as the entry key to the returning object and the channel as a list, which is how it's documented in `napalm-base` right now. Unless I am misinterpreting something.

**Comment by narJH27** [comment]
`show controllers dwdm <interface number> optic` is the command on XR. There is no global command on XR that gives details about all the optics at once, like in JUNOS or EOS.

To answer your question about the interface name, yes we get the interface name for sure. However, since the data structure returned by `get_optics()` is a multi-level dict with the interface name being the top level key, I wasnt sure how to model the test such that I can validate the interface name too. Any thoughts?

**Comment by mirceaulinic** [comment]
> Any thoughts?

Yes, I think you should leave the output unchanged. Adapt only the example and the tests - currently you output is designed to have a depth of 6 levels, while in your tests you go up to 5 levels - you missed something:


```python
>>> get_optics = {
...   'et1': {
...     0: {
...         0: [
...             {
...               'index': 0,
...               'state': {
...                   'input_power': {
...                       'instant': 0.0,
...                       'avg': 0.0,
...                       'min': 0.0,
...                       'max': 0.0,
...                   },
...                   'output_power': {
...                       'instant': 0.0,
...                       'avg': 0.0,
...                       'min': 0.0,
...                       'max': 0.0,
...                   },
...                   'laser_bias_current': {
...                       'instant': 0.0,
...                       'avg': 0.0,
...                       'min': 0.0,
...                       'max': 0.0,
...                   }
...               }
...             }
...         ]
...     }
...   }
... }
>>>
>>> for intf, intf_optic in get_optics.iteritems():
...    for phy_chan in intf_optic:
...        for chan in intf_optic[phy_chan]:
...            for item in intf_optic[phy_chan][chan]:
...                for state, state_vals in (
...                        intf_optic[phy_chan][chan][item].iteritems()):
...                             print state_vals
...
Traceback (most recent call last):
  File "<stdin>", line 6, in <module>
TypeError: list indices must be integers, not dict
```

**Comment by mirceaulinic** [comment]
You probably wanted to do this:

```python
>>> for intf, intf_optic in get_optics.iteritems():
...   for phy_chan, chan in intf_optic.iteritems():
...     for stuff, more_stuff in chan.iteritems():
...       for item in more_stuff:
...          for state, state_vals in item.iteritems():
...             print state_vals
...
0
{'output_power': {'max': 0.0, 'avg': 0.0, 'instant': 0.0, 'min': 0.0}, 'laser_bias_current': {'max': 0.0, 'avg': 0.0, 'instant': 0.0, 'min': 0.0}, 'input_power': {'max': 0.0, 'avg': 0.0, 'instant': 0.0, 'min': 0.0}}
```

**Comment by mirceaulinic** [comment]
@narJH27 

> show controllers dwdm <interface number> optic is the command on XR. There is no global command on XR that gives details about all the optics at once, like in JUNOS or EOS.

Please do not use CLI commands where there's XML request equivalent.

Login into your IOS-XR box, and type: ```xml echo format``` then paste the following lines:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Request MajorVersion="1" MinorVersion="0">
  <Get><Operational><DWDM/></Operational></Get>
</Request>
```

File: napalm_base/test/base.py [hunk 2]
```diff
@@ -473,14 +473,13 @@ def test_get_optics(self):
...
+        result = isinstance(get_optics, list)
 
         for intf, intf_optic in get_optics.iteritems():
-            for phy_chann in intf_optic:
-                for chan in intf_optic[phy_chan]:
-                    for item in intf_optic[phy_chan][chan]:
-                        for state, state_vals in (
-                                intf_optic[phy_chan][chan][item].iteritems()):
+            for phy_chan, chan in intf_optic.iteritems():
+                for stuff, more_stuff in chan.iteritems():
```
Comment (mirceaulinic) [inline napalm_base/test/base.py]: Please do not leave ```stuff``` and ```more_stuff``` :)

**Comment by narJH27** [comment]
Yep will use the XML equivalents for IOSXR.

**Comment by mirceaulinic** [comment]
Only need to update the Example -- see my comment in https://github.com/napalm-automation/napalm-base/pull/58/files#r71615447 then rebase everything in one commit and we're done here

**Comment by narJH27** [comment]
@dbarrosop could you please merge this PR when you get a chance? I am waiting on this so that I can push my changes to the EOS driver, so that the build will not fail

**Comment by dbarrosop** [comment]
@mirceaulinic the output you are using in your example is wrong. The example in the dosctring is correct:

https://github.com/napalm-automation/napalm-base/blob/develop/napalm_base/base.py#L1468

File: napalm_base/base.py [hunk 3]
```diff
@@ -1469,6 +1469,8 @@ def get_optics(self):
 
             {
                     'et1': {
+                        0: {
```
Comment (dbarrosop) [inline napalm_base/base.py]: What is this? This and next line shouldn't be here.

**Comment by mirceaulinic** [comment]
@dbarrosop 
> The example in the dosctring is correct:

Does ```'physical_channels'``` or ```'channel'``` look like valid ```int``` as specified few lines above? https://github.com/napalm-automation/napalm-base/blob/develop/napalm_base/base.py#L1448

File: napalm_base/base.py (hunk 3 above)
Comment (mirceaulinic) [inline napalm_base/base.py]: It'a valid example for the ```int``` field ```physical_channels``` maybe?

**Comment by narJH27** [comment]
@dbarrosop Not sure what you mean by, 'do the `physical_channels` and `channel` look like int'? The example just lists the default value of phy_chan and chan for an interface. Can you explain?

**Comment by dbarrosop** [comment]
@mirceaulinic, you are right, the description is not clear enough but the example is good. The docstring could be somehting like:

```
        Returns a dictionary where the keys are as listed below:
            * intf_name (unicode)
                * physical_channels (key_name)
                    * channels (list of dicts)
                        * index (int)
                        * state
                            * input_power
                                * instant (float)
                                * avg (float)
                                * min (float)
                                * max (float)
                            * output_power
                                * instant (float)
                                * avg (float)
                                * min (float)
                                * max (float)
                            * laser_bias_current
                                * instant (float)
                                * avg (float)
                                * min (float)
                                * max (float)
        Example:
            {
                    'et1': {
                        'physical_channels': {
                            'channel': [
                                {
                                    'index': 0,
                                    'state': {
                                        'input_power': {
                                            'instant': 0.0,
                                            'avg': 0.0,
                                            'min': 0.0,
                                            'max': 0.0,
                                        },
                                        'output_power': {
                                            'instant': 0.0,
                                            'avg': 0.0,
                                            'min': 0.0,
                                            'max': 0.0,
                                        },
                                        'laser_bias_current': {
                                            'instant': 0.0,
                                            'avg': 0.0,
                                            'min': 0.0,
                                            'max': 0.0,
                                        },
                                    }
                                }
                            ]
                        }
                    }
                }
            }
```

Does that make more sense?

**Comment by narJH27** [comment]
@dbarrosop So in this example, what would the value for `physical_channels` and `channel` be in the default case? will it take a value of 0?

**Comment by dbarrosop** [comment]
It's actually hard to describe, we should use YANG xD

`physical_channels` is a key name which returns a dictionary.
`channel` is a key name as well that returns a list of dicts

If we think in OOP `physical_channels` and `channel` would be attributes:

```
result['et1'].physical_channels.channel[0].index # this would return 0
result['et1'].physical_channels.channel[0].state.input_power. instant # this would return 0.0
```

Does this make more sense?

**Comment by mirceaulinic** [comment]
@dbarrosop Yesterday, comparing the structure and the example, I thought the same. But afterwards, after comparing with the implementation in https://github.com/napalm-automation/napalm-eos/pull/32, seemed that the fields in subject are ```int```.
I would like to as @narJH27 to make it clear if possible. Can you please paste, for example the output from your eos implementation to seem what you actually had in mind? Thanks!

**Comment by dbarrosop** [comment]
Maybe we should just write a class to make it more clear and testable?

**Comment by narJH27** [comment]
@mirceaulinic My implementation currently uses '0' as the default value for these, thereby making them ints.

However, since @dbarrosop mentioned that this will be modeled off of OC, I am unsure how the keys will have to be managed in the scheme of this returned data structure. @dbarrosop when you say, write a class and make it testable, what exactly do you mean?

**Comment by dbarrosop** [comment]
I will send later today some code that I think it will help

**Comment by dbarrosop** [comment]
So I rewrote your test:

```
result = device.get_optics()
for iface, iface_data in result.iteritems():
    assert isinstance(iface, str)
    for channel in iface_data['physical_channels']['channel']:
        assert len(channel) == 2
        assert isinstance(channel['index'], int)
        for field in ['input_power', 'output_power', 'laser_bias_current']:
            assert len(channel['state'][field]) ==  4
            assert isinstance(channel['state'][field]['instant'], float)
            assert isinstance(channel['state'][field]['avg'], float)
            assert isinstance(channel['state'][field]['min'], float)
            assert isinstance(channel['state'][field]['max'], float)
```

I hope this helps clarifying the structure.

**Comment by narJH27** [comment]
Just wanted to confirm if the resulting data structure will be a list of dicts, or a dict of dicts?

**Comment by dbarrosop** [comment]
It's a dict where the key is the interface name. If you try a list of dicts with the test I pasted you in my previous comment you will get an error.

**Comment by dbarrosop** [comment]
There is a problem with your current tests and is that it's easy to forge a dictionary that will pass the test without actually being compliant. Could you remove the model you previously added from the `models.py` and replace the test with the code of my previous comment? That code should deal with "rogue" keys as well : )

Thanks!

**Comment by narJH27** [comment]
Done.

**Comment by narJH27** [comment]
@mirceaulinic @dbarrosop  Quick question for you guys! 

I am trying to retrieve the light levels for IOSXR using `<Get><Operational><DWDM/></Operational></Get>`.

However, the output of the above rpc-command, screws up the interface names to dwdm0/0/0/0 etc and so on. I would actually like to retrieve the output of the command `show controllers <intf-name> phy` instead. I have tried scouting through the XML schemas guide for XR and have not found anything that can help out here. Any thoughts on what can be done as an alternative?

**Pull request merged**